python/lsst/meas/base/recalibrateExposure.py

```diff
@@ -0,0 +1,321 @@
+        doc="Photometric calibration",
+        name="{calibSource}_photoCalib",
+        storageClass="PhotoCalib",
+        dimensions=["instrument", "visit", "detector", "tract"],
+        multiple=True,
+    )
+    skyWcs = cT.PrerequisiteInput(
+        doc="Astrometric calibration",
+        name="{calibSource}_skyWcs",
+        storageClass="SkyWcs",
+        dimensions=["instrument", "visit", "detector", "tract"],
+        multiple=True,
+    )
+    skyCorr = cT.PrerequisiteInput(
+        doc="Sky correction",
+        name="skyCorr",
+        storageClass="Background",
+        dimensions=["instrument", "visit", "detector"],
+        multiple=True,
+    )
+
+    def __init__(self, *args, recalibrate="recalibrate", **kwargs):
+        """Initialise
+
+        Filters connections based on run-time task configuration.
+
+        Parameters
+        ----------
+        recalibrate : `str`
+            Name of the `RecalibrateExposureTask` in the ``config``.
+        *args, **kwargs
+            Usual construction arguments for `PipelineTaskConnections`.
+        """
+        super().__init__(*args, **kwargs)
+        config = getattr(self.config, recalibrate)
+        if not config.doApplyExternalPhotoCalib:
+            self.prerequisiteInputs.discard("photoCalib")
+        if not config.doApplyExternalSkyWcs:
+            self.prerequisiteInputs.discard("skyWcs")
+        if not config.doApplySkyCorr:
+            self.prerequisiteInputs.discard("skyCorr")
+
+
+class RecalibrateExposureTask(Task):
+    """Read and apply calibrations to a exposure
+
+    This is used to apply calibrations from jointcal or fgcm to a exposure.
+
+    If ``config.doApplyExternalPhotoCalib`` is ``True``, the photometric
+    calibration (``photoCalib``) is taken from ``config.externalPhotoCalibName``
+    via the ``<name>_photoCalib`` dataset.  Otherwise, the photometric
+    calibration is retrieved from the processed exposure.
+
+    When ``config.doApplyExternalSkyWcs` is ``True``, the astrometric
+    calibration is taken from ``config.externalSkyWcsName`` with the
+    ``<name>_wcs`` dataset. Otherwise, the astrometric calibration is taken from
+    the processed exposure.
+
+    If ``config.doApplySkyCorr`` is ``True``, the sky background is read from
+    dataset ``skyCorr``.
+
+    Use in Gen3
+    -----------
+
+        class FooConfig(lsst.pex.config.Config):
+            recalibrate = lsst.pex.config.ConfigurableField(
+                target=RecalibrateExposureTask,
+                doc="Read calibrated exposure"
+            )
+
+        class FooTaskConnections(RecalibrateExposureConnections):
+            ...
+
+        class FooTask(lsst.pipe.base.PipelineTask):
+            def runQuantum(self, butlerQC, inputRefs, outputRefs):
+                inputs = butlerQC.get(inputRefs)
+                calibs = self.recalibrate.extractCalibs(inputs, len(inputs["exposure"]))
+                self.recalibrate.runMultiple(inputs["exposure"], **calibs)
```

Comment:
Instead of providing `runMultiple`, I think I'd recommend that `runQuantum` implementors just call `run` as many times as they need, on the explicit exposure-like datasets they actually have.  Making `runQuantum` less magic at the expense of a bit of boilerplate is a net win, especially because most of the time there will only be a single call to `run`.